### PR TITLE
workload balancer and datasets folder refactor

### PR DESCRIPTION
**Pull request by JacoCheung**
## Description
Address #207 .

This PR adds a batch shuffler that can balance the workloads among data patallel group. User can set `TrainerArgs.enable_balanced_shuffler = True` to enable this feature. It does purely gemm compute workloads.  

## Breaking changes:
1. RetrievalBatch and RankingBatch are consolidated into one. They all subclass `BaseBatch`
The BaseBatch definition. In the future, a model-specific batch should subclass this class.

```
@dataclass
class BaseBatch(Pipelineable):
    """
    All tensors must share a same batch size.
    """

    features: KeyedJaggedTensor
    batch_size: int  # local batch size
    feature_to_max_seqlen: Dict[str, int]

    contextual_feature_names: List[str] = field(default_factory=list)
    # when labels is a tensor, it means the labels can be reshaped to [actual_batch_size, ...] and select along the batch dimension.
    labels: Union[KeyedJaggedTensor, torch.Tensor] = None
    actual_batch_size: Optional[int] = None  # in case of padding.
```
2. The labels in hstu Batch is changed into KeyJaggedTensor for easier shuffling. 

3. Batch shuffler is added into train pipeline, follows H2D. 

4. The hstu dataset args `max_sequence_length` is now `max_history_seqlen`, the full seqlen should be `max_history_seqlen * 2 + max_num_candidates * 2 + num_contextual_features`.

5. Datasets are moved under commons.

~~[CI](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/42614598)~~ 
~~[CI](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/42630027)~~
~~[CI](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/42697712)~~
~~[CI](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/42714340)~~
~~[CI](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/43157099)~~
[CI](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/43249164)

### CODE REVIEW

File: examples/commons/sequence_batch/batch.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            weights = output[2] if len(output) > 2 else None
+            return KeyedJaggedTensor.from_lengths_sync(
+                keys=features.keys(), values=values, lengths=lengths, weights=weights
+            )
+
+        if self.actual_batch_size != self.batch_size:
+            # we use the actual batch size to filter the indices.
+            actual_batch_size = indices[indices < self.actual_batch_size].numel()
+        else:
+            actual_batch_size = indices.numel()
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/sequence_batch/batch.py:141]: I've verified, thanks!

File: examples/commons/distributed/batch_shuffler.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        # 2. Partition the workloads
+        partitions_indices = karmarkar_karp(
+            allgather_workloads, num_partitions, equal_size=True
+        )
+        indices_this_rank = torch.tensor(
+            partitions_indices[rank],
+            dtype=torch.int64,
+            device=batch.features.lengths().device,
+        )
+        indices_this_rank, _ = torch.sort(indices_this_rank)  #
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py:50]: Yeah, it's necessary, since we need to guarantee that the padding values are contiguously stored at the tail.
I've added some comments. The batchsize is not that large, so overhead is acceptable.

File: examples/hstu/training/README.md [hunk 3]
```diff
@@ -70,11 +70,11 @@ Before getting started, please make sure that all pre-requisites are fulfilled.
...
 ### Dataset preprocessing
 
-In order to prepare the dataset for training, you can use our `preprocessor.py` under the hstu example folder of the project.
+In order to prepare the dataset for training, you can use our `hstu_data_preprocessor.py` under the commons folder of the project.
 
 ```bash
-cd <root-to-repo>/examples/hstu && 
-mkdir -p ./tmp_data && python3 ./preprocessor.py --dataset_name <"ml-1m"|"ml-20m"|"kuairand-pure"|"kuairand-1k"|"kuairand-27k">
+cd <root-to-repo>/examples/commons && 
+mkdir -p ./tmp_data && python3 ./hstu_data_preprocessor.py --dataset_name <"ml-1m"|"ml-20m"|"kuairand-pure"|"kuairand-1k"|"kuairand-27k">
```
Comment (shijieliu) [inline examples/hstu/training/README.md:77]: is this needed? Can we make `data_preprocessor.py` as before

File: examples/tests/commons/test_batch.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
+import sys
```
Comment (shijieliu) [inline examples/tests/commons/test_batch.py:1]: do we add the corresponding part in CI

File: examples/hstu/training/README.md (hunk 3 above)
Comment (JacoCheung) [reply examples/hstu/training/README.md:77]: I'm not sure. Now we have sid and hstu datasets, I assume they will not share the preprocessor.( Though we don't have a sid preprocessor yet). 

Or we can consolidte them into one?

File: examples/tests/commons/test_batch.py (hunk 4 above)
Comment (JacoCheung) [reply examples/tests/commons/test_batch.py:1]: Yes, I did create an MR. See the 2nd CI line above in the desription.

File: examples/hstu/modules/hstu_processor.py [hunk 5]
```diff
@@ -95,6 +95,11 @@ def hstu_preprocess_embeddings(
             )
             sequence_max_seqlen = sequence_max_seqlen * 2
         else:
+            # TODO@junyi: We can optimize the concat:
+            # 1. use jagged split to get [history_embs, candidate_embs]
+            # 2. use cat to interleave the history_embs and history_action_embs part
+            # 3. use jagged concat to append the candidate_embs
+
```
Comment (JacoCheung) [inline examples/hstu/modules/hstu_processor.py:102]: I happened to see the for loop concat. And I think there should be a chance to optimize with the steps I commented above.  Just for a reference @geoffreyQiu

File: examples/commons/sequence_batch/batch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/sequence_batch/batch.py:141]: No. There isn't `_global_actual_batch_size` field. This function is invoked after allgather, so `self.actual_batch_size
` is already the global actual batch size.

File: examples/commons/distributed/batch_shuffler.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        Tuple[BaseBatch, torch.Tensor, torch.Tensor],
+    ]:
+        workloads = self.get_workloads(batch, *args, **kwargs)
+        assert (
+            workloads.shape[0] == batch.batch_size
+        ), "workloads should have the same shape as batch_size"
+        num_partitions = torch.distributed.get_world_size(pg_group)
+        rank = torch.distributed.get_rank(pg_group)
+        # 1. Allgather the workloads
+        allgather_workloads = gather_along_first_dim(workloads, pg_group)
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py:37]: Yes, the seqlen for padded batch is set to 0 and corresponding workload is zero.

File: examples/commons/perf_model/partitioner.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        workloads.tolist()
+        if isinstance(workloads, Tensor) and Tensor is not None
+        else workloads
+    )
+    sorted_workloads = sorted([(workload, i) for i, workload in enumerate(workloads)])
+    states_pq: List[Any] = []
+    if equal_size:
+        assert (
+            len(workloads) % k_partitions == 0
+        ), f"{len(workloads)} % {k_partitions} != 0"
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/perf_model/partitioner.py:125]: The incomplete batch is padded with zero workload. Assertion is successful.

File: examples/commons/pipeline/train_pipeline.py [hunk 8]
```diff
@@ -445,11 +393,13 @@ def copy_batch_to_gpu(
...
         """
         context = self._create_context()
-        with nvtx.annotate(f"## copy_batch_to_gpu {self._next_index} ##"):
+        with nvtx.annotate(f"## copy_batch_to_gpu_and_shuffle {self._next_index} ##"):
             with self._stream_context(self._memcpy_stream):
                 batch = self._next_batch(dataloader_iter)
                 if batch is not None:
                     batch = _to_device(batch, self._device, non_blocking=True)
+                    # TODO@junzhang, there are cpu ops / nccl comm and lots of sync in shuffle.
+                    batch = self._batch_shuffle(batch)
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/pipeline/train_pipeline.py:402]: Will try to optimize the shuffle in the future.

File: examples/hstu/modules/hstu_block.py [hunk 9]
```diff
@@ -56,14 +56,14 @@ def __init__(
     def forward(
         self,
         embeddings: Dict[str, JaggedTensor],
-        batch: Union[RankingBatch, RetrievalBatch],
+        batch: Union[HSTUBatch, HSTUBatch],
```
Comment (shijieliu) [inline examples/hstu/modules/hstu_block.py]: need fix.

File: examples/commons/sequence_batch/batch.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    For incomplete batch, all dense tensors and KJT offsets/lengths are padded to the batch size.
+    """
+
+    features: KeyedJaggedTensor
+    batch_size: int  # local batch size
+    feature_to_max_seqlen: Dict[str, int]
+
+    contextual_feature_names: List[str] = field(default_factory=list)
+    # when labels is a tensor, it means the labels can be reshaped to [actual_batch_size, ...] and select along the batch dimension.
+    labels: Union[KeyedJaggedTensor, torch.Tensor] = None
```
Comment (shijieliu) [inline examples/commons/sequence_batch/batch.py]: could you tell me when the labels will be a tensor instead of a KJT?

File: examples/hstu/inference/README.md [hunk 11]
```diff
@@ -97,7 +97,7 @@ Turn on option `INFERENCEBUILD=1` to skip Megatron installation, which is not re
 ~$ export PYTHONPATH=${PYTHONPATH}:$(realpath ../)
 ~$ 
 ~$ # Proprocess the dataset for inference:
-~$ python3 ./preprocessor.py --dataset_name "kuairand-1k" --inference
+~$ python3 ../commons/hstu_data_preprocessor.py --dataset_name "kuairand-1k" --inference
```
Comment (shijieliu) [inline examples/hstu/inference/README.md:100]: why do we move preprocessor.py into commons?

File: examples/commons/sequence_batch/batch.py (hunk 10 above)
Comment (JacoCheung) [reply examples/commons/sequence_batch/batch.py]: Currently no cases. 

The labels for SID GR used to be a dense tensor . However, it's not convenient for incomplete batch shuffling, so I changed the original output label to kjt with padded lengths. 
I would like to reserve the dense option, though.

File: examples/hstu/inference/README.md (hunk 11 above)
Comment (JacoCheung) [reply examples/hstu/inference/README.md:100]: Because `commons/datasets/hstu_sequence_dataset` [depends on](https://github.com/JacoCheung/recsys-examples/blob/junzhang/workload_balancer/examples/commons/datasets/hstu_sequence_dataset.py#L38)  this file. Otherwise we would have something like 
```
from examples.hstu.preprocessor import get_common_preprocessors
```
under `commons` folder. I would assume commons should not rely on the more specific folder. (it should only rely on commons/*)

Comment (shijieliu) [reply examples/hstu/inference/README.md:100]: gotcha

File: examples/commons/sequence_batch/batch.py (hunk 10 above)
Comment (shijieliu) [reply examples/commons/sequence_batch/batch.py]: but if you reserve the dense option, it's not enough to just mark here. You need to do the check in everywhere you use it. That makes the dense option workable.

Comment (JacoCheung) [reply examples/commons/sequence_batch/batch.py]: I see. I fixed to KJT.

**Review by shijieliu — APPROVED** [review]
